File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/model/BaseDraw.java
package pisarev.com.modeling.mvp.model;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;

import java.util.ArrayList;

import javax.inject.Inject;

import pisarev.com.modeling.application.App;
import pisarev.com.modeling.interfaces.ViewMvp;

public abstract class BaseDraw {

    private Paint paintFullLine;
    private Paint paintDottedLine;
    boolean clockwise;
    Paint line;
    ArrayList<StringBuffer> programList;
    ArrayList<Frame> frameList;
    ViewMvp.MyViewMvp myViewMvp;
    @Inject
    MyData data;

    BaseDraw(ViewMvp.MyViewMvp myViewMvp) {
        this.myViewMvp = myViewMvp;
        init();
    }

    private void init() {
        App.getComponent().inject( this );
        programList = data.getProgramList();
        frameList =data.getFrameList();
        line = new Paint();
        paintFullLine = new Paint();
        paintFullLine.setColor( Color.GREEN );
        paintFullLine.setStyle( Paint.Style.STROKE );
        paintFullLine.setAntiAlias( true );
        paintDottedLine = new Paint();
        paintDottedLine.setColor( Color.GRAY );
        paintDottedLine.setStyle( Paint.Style.STROKE );
        paintDottedLine.setAntiAlias( true );
        paintDottedLine.setPathEffect( new DashPathEffect( new float[]{20f, 10f}, 0f ) );
    }

    public abstract void drawContour(Canvas canvas, Point pointCoordinateZero, float zoom, int index);

    void drawLine(Canvas canvas, Paint paint, Point pointCoordinateZero, Point pointStart, Point pointEnd, float zoom) {
        Path path = new Path();
        Point pStart = new Point( pointStart.getX(), pointStart.getZ() );
        Point pEnd = new Point( pointEnd.getX(), pointEnd.getZ() );
        pStart.setX( pStart.getX() * zoom );
        pStart.setZ( pStart.getZ() * zoom );
        pEnd.setX( pEnd.getX() * zoom );
        pEnd.setZ( pEnd.getZ() * zoom );
        if (pStart.getZ() > 0) pStart.setZ( pointCoordinateZero.getZ() - pStart.getZ() );
        else pStart.setZ( pointCoordinateZero.getZ() + Math.abs( pStart.getZ() ) );
        if (pEnd.getZ() > 0) pEnd.setZ( pointCoordinateZero.getZ() - pEnd.getZ() );
        else pEnd.setZ( pointCoordinateZero.getZ() + Math.abs( pEnd.getZ() ) );
        path.moveTo( pointCoordinateZero.getX() + pStart.getX(), pStart.getZ() );
        path.lineTo( pointCoordinateZero.getX() + pEnd.getX(), pEnd.getZ() );
        canvas.drawPath( path, paint );
    }

    void drawArc(Canvas canvas, Paint paint, Point pointCoordinateZero, Point pointStart, Point pointEnd, float radius, float zoom, boolean clockwise) {
        Path path = new Path();
        Point pStart = new Point( pointStart.getX(), pointStart.getZ() );
        Point pEnd = new Point( pointEnd.getX(), pointEnd.getZ() );
        RectF rectF = new RectF();
        pStart.setX( pStart.getX() * zoom );
        pStart.setZ( pStart.getZ() * zoom );
        pEnd.setX( pEnd.getX() * zoom );
        pEnd.setZ( pEnd.getZ() * zoom );
        radius *= zoom;
        float startAngle = 0, sweetAngle, cathetus;
        float chord = (float) Math.sqrt( Math.pow( pStart.getX() - pEnd.getX(), 2 ) + Math.pow( pStart.getZ() - pEnd.getZ(), 2 ) );
        float h = (float) Math.sqrt( radius * radius - (chord / 2) * (chord / 2) );
        if (clockwise) {
            float x01 = (pStart.getX() + (pEnd.getX() - pStart.getX()) / 2 + h * (pEnd.getZ() - pStart.getZ()) / chord);
            float z01 = (pStart.getZ() + (pEnd.getZ() - pStart.getZ()) / 2 - h * (pEnd.getX() - pStart.getX()) / chord);
            if (pStart.getX() > x01 && pStart.getZ() >= z01) {
                cathetus = pStart.getX() - x01;
                if (pStart.getZ() == z01)
                    startAngle = 0;
                else startAngle = (float) (360 - Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            if (pStart.getX() >= x01 && pStart.getZ() < z01) {
                cathetus = pStart.getX() - x01;
                if (pStart.getX() == x01)
                    startAngle = 90;
                else startAngle = (float) (Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            if (pStart.getX() < x01 && pStart.getZ() <= z01) {
                cathetus = x01 - pStart.getX();
                if (pStart.getZ() == z01)
                    startAngle = 180;
                else startAngle = (float) (180 - Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            if (pStart.getX() <= x01 && pStart.getZ() > z01) {
                cathetus = x01 - pStart.getX();
                if (pStart.getX() == x01)
                    startAngle = 270;
                else startAngle = (float) (180 + Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            rectF.set( pointCoordinateZero.getX() + x01 - radius, pointCoordinateZero.getZ() - z01 - radius, pointCoordinateZero.getX() + x01 + radius, pointCoordinateZero.getZ() - z01 + radius );
        } else {
            float x02 = pStart.getX() + (pEnd.getX() - pStart.getX()) / 2 - h * (pEnd.getZ() - pStart.getZ()) / chord;
            float z02 = pStart.getZ() + (pEnd.getZ() - pStart.getZ()) / 2 + h * (pEnd.getX() - pStart.getX()) / chord;
            if (pEnd.getX() > x02 && pEnd.getZ() >= z02) {
                cathetus = pEnd.getX() - x02;
                if (pEnd.getZ() == z02)
                    startAngle = 0;
                else startAngle = (float) (360 - Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            if (pEnd.getX() >= x02 && pEnd.getZ() < z02) {
                cathetus = pEnd.getX() - x02;
                if (pEnd.getX() == x02)
                    startAngle = 90;
                else startAngle = (float) (Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            if (pEnd.getX() < x02 && pEnd.getZ() <= z02) {
                cathetus = x02 - pEnd.getX();
                if (pEnd.getZ() == z02)
                    startAngle = 180;
                else startAngle = (float) (180 - Math.acos( cathetus / radius ) * (180 / Math.PI));
            }

            if (pEnd.getX() <= x02 && pEnd.getZ() > z02) {
                cathetus = x02 - pEnd.getX();
                if (pEnd.getX() == x02)
                    startAngle = 270;
                else startAngle = (float) (180 + Math.acos( cathetus / radius ) * (180 / Math.PI));
            }
            rectF.set( pointCoordinateZero.getX() + x02 - radius, pointCoordinateZero.getZ() - z02 - radius, pointCoordinateZero.getX() + x02 + radius, pointCoordinateZero.getZ() - z02 + radius );
        }
        sweetAngle = (float) (2 * Math.asin( chord / (2 * radius) ) * (180 / Math.PI));
        path.addArc( rectF, startAngle, sweetAngle );
        canvas.drawPath( path, paint );
    }

    void drawPoint(Canvas canvas, Point pointCoordinateZero, Point pointEnd, float zoom) {
        float radiusPoint = 7F;
        Paint paint = new Paint();
        paint.setStyle( Paint.Style.FILL );
        paint.setColor( Color.RED );
        Path path = new Path();
        Point pEnd = new Point( pointEnd.getX(), pointEnd.getZ() );
        pEnd.setX( pEnd.getX() * zoom );
        pEnd.setZ( pEnd.getZ() * zoom );
        if (pEnd.getZ() > 0) pEnd.setZ( pointCoordinateZero.getZ() - pEnd.getZ() );
        else pEnd.setZ( pointCoordinateZero.getZ() + Math.abs( pEnd.getZ() ) );
        path.addCircle( pointCoordinateZero.getX() + pEnd.getX(), pEnd.getZ(), radiusPoint, Path.Direction.CW );
        canvas.drawPath( path, paint );
    }

    private boolean isG17(ArrayList<StringBuffer> programList) {
        for (int i = 0; i < programList.size(); i++)
            if (programList.get( i ).toString().contains( "G17" )) {
                return true;
            }
        return false;
    }

    void checkGCode(ArrayList<String>gCodeList){
        boolean isG17 = isG17( programList );
        for (String gCode:gCodeList) {
            switch (gCode) {
                case "G0":
                case "G00":
                    line = paintDottedLine;
                    break;
                case "G1":
                case "G01":
                    line = paintFullLine;
                    break;
                case "G2":
                case "G02":
                    clockwise = isG17;
                    break;
                case "G3":
                case "G03":
                    clockwise = !isG17;
                    break;
            }
        }
    }
}

File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/contur/base/BaseDraw.kt
package pisarev.com.modeling.mvp.contur.base

import android.graphics.*
import pisarev.com.modeling.interfaces.IDraw
import pisarev.com.modeling.mvp.contur.Point
import pisarev.com.modeling.mvp.contur.Point2D
import pisarev.com.modeling.mvp.model.core.GCode
import kotlin.math.acos
import kotlin.math.pow
import kotlin.math.sqrt

abstract class BaseDraw : GCode {
    protected var draw: IDraw? = null
    protected var isNumberLine = false
    protected var numberLine = 0
    protected var colorPoint = Color.RED
    protected var colorTouchPoint = Color.rgb(0, 128, 255)
    private var line: Paint? = null
    private var paintFullLine: Paint? = null
    private var paintDottedLine: Paint? = null

    protected constructor(draw: IDraw?) {
        this.draw = draw
        init()
    }

    private fun init() {
        line = Paint()
        paintFullLine = Paint()
        paintFullLine!!.color = Color.GREEN
        paintFullLine!!.style = Paint.Style.STROKE
        paintFullLine!!.isAntiAlias = true
        paintDottedLine = Paint()
        paintDottedLine!!.color = Color.GRAY
        paintDottedLine!!.style = Paint.Style.STROKE
        paintDottedLine!!.isAntiAlias = true
        paintDottedLine!!.pathEffect = DashPathEffect(floatArrayOf(12f, 7f), 0f)
    }

    protected fun drawLine(
        canvas: Canvas,
        isRapidFeed: Int,
        pointCoordinateZero: Point,
        pointStart: Point,
        pointEnd: Point,
        zoom: Float
    ) {
        when(isRapidFeed){
            0 -> line = paintDottedLine
            1 -> line =  paintFullLine
        }
        val path = Path()
        val pStart = Point(pointStart)
        val pEnd = Point(pointEnd)
        val pointCoordinateZeroF = Point(pointCoordinateZero)
        pStart.x = pStart.x * zoom
        pStart.z = pStart.z * zoom
        pEnd.x = pEnd.x * zoom
        pEnd.z = pEnd.z * zoom
        if (pStart.z > 0) pStart.z = pointCoordinateZeroF.z - pStart.z else pStart.z =
            pointCoordinateZeroF.z + Math.abs(pStart.z)
        if (pEnd.z > 0) pEnd.z = pointCoordinateZeroF.z - pEnd.z else pEnd.z = pointCoordinateZeroF.z + Math.abs(pEnd.z)

        path.moveTo(pointCoordinateZeroF.x + pStart.x, pStart.z)
        path.lineTo(pointCoordinateZeroF.x + pEnd.x, pEnd.z)
        canvas.drawPath(path, line!!)
    }

    protected fun drawArc(
        canvas: Canvas,
        isRapidFeed: Int,
        pointCoordinateZero: Point,
        pointStart: Point,
        pointEnd: Point,
        radius: Float,
        zoom: Float,
        clockwise: Int
    ) {
        when(isRapidFeed){
            0 -> line = paintDottedLine
            1 -> line =  paintFullLine
        }
        var r = radius
        val path = Path()
        val pStart = Point(pointStart)
        val pEnd = Point(pointEnd)
        val pointCoordinateZeroF = Point(pointCoordinateZero)
        val rectF = RectF()
        pStart.x = pStart.x * zoom
        pStart.z = pStart.z * zoom
        pEnd.x = pEnd.x * zoom
        pEnd.z = pEnd.z * zoom
        r *= zoom
        var startAngle = 0f
        val sweetAngle: Float
        var cathetus: Float
        val chord = sqrt(
            (pStart.x - pEnd.x).toDouble().pow(2.0) + (pStart.z - pEnd.z).toDouble().pow(2.0)
        ).toFloat()
        val h = sqrt((r * r - chord / 2 * (chord / 2)).toDouble()).toFloat()
        if (clockwise == 2) {
            val x01: Float =
                pStart.x + (pEnd.x - pStart.x) / 2 + h * (pEnd.z - pStart.z) / chord
            val z01: Float =
                pStart.z + (pEnd.z - pStart.z) / 2 - h * (pEnd.x - pStart.x) / chord
            if (pStart.x > x01 && pStart.z >= z01) {
                cathetus = pStart.x - x01
                startAngle =
                    if (pStart.z == z01) 0f else (360 - acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pStart.x >= x01 && pStart.z < z01) {
                cathetus = pStart.x - x01
                startAngle =
                    if (pStart.x == x01) 90f else (acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pStart.x < x01 && pStart.z <= z01) {
                cathetus = x01 - pStart.x
                startAngle =
                    if (pStart.z == z01) 180f else (180 - acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pStart.x <= x01 && pStart.z > z01) {
                cathetus = x01 - pStart.x
                startAngle =
                    if (pStart.x == x01) 270f else (180 + acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            rectF[pointCoordinateZeroF.x + x01 - r, pointCoordinateZeroF.z - z01 - r, pointCoordinateZeroF.x + x01 + r] =
                pointCoordinateZeroF.z - z01 + r
        }
        if (clockwise == 3) {
            val x02: Float =
                pStart.x + (pEnd.x - pStart.x) / 2 - h * (pEnd.z - pStart.z) / chord
            val z02: Float =
                pStart.z + (pEnd.z - pStart.z) / 2 + h * (pEnd.x - pStart.x) / chord
            if (pEnd.x > x02 && pEnd.z >= z02) {
                cathetus = pEnd.x - x02
                startAngle =
                    if (pEnd.z == z02) 0f else (360 - acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pEnd.x >= x02 && pEnd.z < z02) {
                cathetus = pEnd.x - x02
                startAngle =
                    if (pEnd.x == x02) 90f else (acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pEnd.x < x02 && pEnd.z <= z02) {
                cathetus = x02 - pEnd.x
                startAngle =
                    if (pEnd.z == z02) 180f else (180 - acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            if (pEnd.x <= x02 && pEnd.z > z02) {
                cathetus = x02 - pEnd.x
                startAngle =
                    if (pEnd.x == x02) 270f else (180 + acos((cathetus / r).toDouble()) * (180 / Math.PI)).toFloat()
            }
            rectF[pointCoordinateZeroF.x + x02 - r, pointCoordinateZeroF.z - z02 - r, pointCoordinateZeroF.x + x02 + r] =
                pointCoordinateZeroF.z - z02 + r
        }
        sweetAngle = (2 * Math.asin((chord / (2 * r)).toDouble()) * (180 / Math.PI)).toFloat()
        path.addArc(rectF, startAngle, sweetAngle)
        canvas.drawPath(path, line!!)
    }

    protected fun drawRND(
        canvas: Canvas,
        isRapidFeed: Int,
        pointSystemCoordinate: Point,
        pointStart: Point,
        pointEnd: Point,
        pointF: Point,
        radiusRND: Float,
        zoom: Float
    ) {
        when(isRapidFeed){
            0 -> line = paintDottedLine
            1 -> line =  paintFullLine
        }
        val pStart = Point(pointStart)
        val pEnd = Point(pointEnd)
        val pointStartCR = Point()
        val pointEndCR = Point()
        val pointF = Point(pointF)
        val cathet: Float
        var clockwiseRND = 3
        val angle = Point2D(pEnd.x - pStart.x, pEnd.z - pStart.z).angle(pEnd.x - pointF.x, pEnd.z) - pointF.z
        val firstDistance = Point2D(pStart.x, pStart.z).distance(pEnd.x, pEnd.z)
        val secondDistance = Point2D(pEnd.x, pEnd.z).distance(pointF.x, pointF.z)
        cathet = if (angle == 90F) {
            radiusRND
        } else {
            ((180 - angle) / 2 * (Math.PI / 180) * radiusRND).toFloat()
        }
        var differenceX = pStart.x - pEnd.x
        var differenceZ = pStart.z - pEnd.z
        pointStartCR.x = differenceX * cathet / firstDistance
        pointStartCR.z = differenceZ * cathet / firstDistance
        pointStartCR.x = pEnd.x + pointStartCR.x
        pointStartCR.z = pEnd.z + pointStartCR.z
        differenceX = pointF.x - pEnd.x
        differenceZ = pointF.z - pEnd.z
        pointEndCR.x = differenceX * cathet / secondDistance
        pointEndCR.z = differenceZ * cathet / secondDistance
        pointEndCR.x = pEnd.x + pointEndCR.x
        pointEndCR.z = pEnd.z + pointEndCR.z
        if (pStart.x > pointF.x && (pStart.z + pointF.z) / 2 > pEnd.z) {
            clockwiseRND = 2
        }
        if (pStart.x > pointF.x && (pStart.z + pointF.z) / 2 < pEnd.z) {
            clockwiseRND = 3
        }
        if (pStart.x < pointF.x && (pStart.z + pointF.z) / 2 < pEnd.z) {
            clockwiseRND = 2
        }
        drawLine(canvas, isRapidFeed, pointSystemCoordinate, Point(pStart), Point(pointStartCR), zoom)
        drawArc(
            canvas,
            isRapidFeed,
            pointSystemCoordinate,
            Point(pointStartCR),
            Point(pointEndCR),
            radiusRND,
            zoom,
            clockwiseRND
        )
        pEnd.x = pointEndCR.x
        pEnd.z = pointEndCR.z
    }

    protected fun drawPoint(canvas: Canvas, pointCoordinateZero: Point, pointEnd: Point, color: Int, zoom: Float) {
        val radiusPoint = 9f
        val paint = Paint()
        paint.style = Paint.Style.FILL
        paint.color = color
        val path = Path()
        val pEnd = Point(pointEnd)
        val pointCoordinateZeroF = Point(pointCoordinateZero)
        pEnd.x = pEnd.x * zoom
        pEnd.z = pEnd.z * zoom
        if (pEnd.z > 0) pEnd.z = pointCoordinateZeroF.z - pEnd.z else pEnd.z = pointCoordinateZeroF.z + Math.abs(pEnd.z)

        path.addCircle(pointCoordinateZeroF.x + pEnd.x, pEnd.z, radiusPoint, Path.Direction.CW)
        canvas.drawPath(path, paint)
    }

    protected fun isG17(gCodes: List<String>): Boolean {
        if(gCodes.toString().contains(G17)) return true
        return false
    }


}